okhttp-urlconnection/src/test/java/okhttp3/OkUrlFactoryTest.java

```diff
@@ -211,6 +213,22 @@ public void checkURLPermitted(URL url) throws IOException {
     }
   }
 
+  @Test
+  public void usesValidHeaderValueForDefaultUserAgent() throws Exception {
```

Comment:
TIOLI: care to add a few more tests while making this change to explain the code?

    System.out.println("🔊".length()); // 2
    System.out.println("🔊".codePointCount(0, 2)); // 1


```
  @Test
  public void usesValidHeaderValueForDefaultUserAgent() throws Exception {
    String userAgent = "🍩" + "\u001F" + '\u007f' + ('\u001f' + 1) + ('\u007f' - 1)
        + "? ♦♥♠♣" + "🍩";
    String expected = "???" + ('\u001f' + 1) + ('\u007f' - 1) + "? ?????";
    testUserAgentEncoding(userAgent, expected);
  }

  @Test
  public void usesStraightHeaderValueForSimpleUserAgent() throws Exception {
    String userAgent = "ABC";
    String expected = "ABC";
    testUserAgentEncoding(userAgent, expected);
  }

  @Test
  public void handlesBadUnicodeStrings() throws Exception {
    String userAgent =  "🔊".substring(0, 1);
    String expected = "?";
    testUserAgentEncoding(userAgent, expected);
  }

  private void testUserAgentEncoding(String userAgent, String expected)
      throws IOException, InterruptedException {
    System.setProperty("http.agent", userAgent);
    server.enqueue(new MockResponse().setResponseCode(200));
    InputStream inputStream = factory.open(server.url("/").url()).getInputStream();
    long skipped;
    do {
      skipped = inputStream.skip(Long.MAX_VALUE);
    } while (skipped != 0);
    RecordedRequest recordedRequest = server.takeRequest();
    assertEquals(expected,
        recordedRequest.getHeader("User-Agent"));
  }
```

Reply:
nice, thanks. will do!